sound/soc/sof/debug-dsp-ops.c

```diff
@@ -0,0 +1,161 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+
+#include <linux/debugfs.h>
+#include <sound/sof/debug.h>
+#include "sof-priv.h"
+#include "ops.h"
+
+/*
+ * set dsp power state op by writing the power state.
+ * ex: echo set_power_state,D3 > dsp_test_op
+ */
+static int sof_dsp_ops_set_power_state(struct snd_sof_dev *sdev, char *state)
+{
+	/* only D3 supported for now */
+	if (strcmp(state, "D3")) {
+		dev_err(sdev->dev, "Unsupported state %s\n", state);
+		return -EINVAL;
+	}
+
+	/* power off the DSP */
+	if (sdev->dsp_power_state.state == SOF_DSP_PM_D0) {
+		const struct sof_ipc_pm_ops *pm_ops = sof_ipc_get_ops(sdev, pm);
+		pm_message_t pm_state;
+		int ret;
+
+		pm_state.event = SOF_DSP_PM_D3;
+
+		/* suspend DMA trace */
+		sof_fw_trace_suspend(sdev, pm_state);
+
+		/* notify DSP of upcoming power down */
+		if (pm_ops && pm_ops->ctx_save) {
+			ret = pm_ops->ctx_save(sdev);
+			if (ret < 0)
+				return ret;
+		}
+
+		ret = snd_sof_dsp_runtime_suspend(sdev);
+		if (ret < 0) {
+			dev_err(sdev->dev, "failed to power off DSP\n");
+			return ret;
+		}
+
+		sdev->enabled_cores_mask = 0;
+		sof_set_fw_state(sdev, SOF_FW_BOOT_NOT_STARTED);
+	}
+
+	return 0;
+}
+
+/*
+ * test firmware boot by passing the firmware file as the argument:
+ * ex: echo boot_firmware,intel/avs/tgl/community/dsp_basefw.bin > dsp_test_op
+ */
+static int sof_dsp_ops_boot_firmware(struct snd_sof_dev *sdev, char *fw_filename)
+{
+	int ret;
+
+	/* power off the DSP */
+	ret = sof_dsp_ops_set_power_state(sdev, "D3");
+	if (ret < 0)
+		return ret;
+
+	if (sdev->basefw.fw)
+		snd_sof_fw_unload(sdev);
+
+	ret = snd_sof_dsp_runtime_resume(sdev);
+	if (ret < 0)
+		return ret;
+
+	sdev->first_boot = true;
+
+	/* load and boot firmware */
+	ret = snd_sof_load_firmware(sdev, (const char *)fw_filename);
+	if (ret < 0)
+		return ret;
+
+	sof_set_fw_state(sdev, SOF_FW_BOOT_IN_PROGRESS);
+
+	ret = snd_sof_run_firmware(sdev);
+	if (ret < 0)
+		return ret;
+
+	/* resume DMA trace */
+	return sof_fw_trace_resume(sdev);
```

Comment:
this looks like code duplication, I don't follow what it's supposed to do

Do we really want to do this or run something like the python stuff on top of a simplified SOF driver.

We have to be really careful not to open a pandora box of new stuff that we'll have to maintain forever twice.

Reply:
@plbossart the idea is to be able to run the python scripts on Linux with minimal effort by using the debugfs option. This is in essence a simplified SOF driver isnt it?

Reply:
So is this basically an alternative to https://github.com/thesofproject/sof-diagnostic-driver that don't have to load/unload?

Isn't the sof-diagnostic-driver approach more generic/flexible?

Reply:
Well, the question is really "what is the partitioning between scripts and driver"? Do we want to reuse snd_sof_run_firmware and hard-code it in a sequence used by a script, or do we want the scripts to do something that runs the firmware?
If you really want to use scripts, then the driver part should be rather minimal and probably a pass-through to access registers.

Reply:
but @plbossart if we want to write scripts to do all the register reads/writes, it would essentially be rewriting the SOF driver. Isn't this a saner approach?

Reply:
> So is this basically an alternative to https://github.com/thesofproject/sof-diagnostic-driver that don't have to load/unload?
> 
> Isn't the sof-diagnostic-driver approach more generic/flexible?

@marc-hb this isnt a replacement for the disgnostic driver. This is a way to enable us to run the python scripts for FW testing on Linux with minimal effort.

Reply:
> but @plbossart if we want to write scripts to do all the register reads/writes, it would essentially be rewriting the SOF driver. Isn't this a saner approach?

the question is whether
a) you want to reuse the existing Python scripts used by FW validation or
b) you want to develop NEW python scripts based on a TBD API

It's not clear from this PR what the direction is. 
a) is complicated because I don't know what the interface requirements are, but 
b) doesn't seem very appealing to me since it'd be a 3rd way of interacting with the DSP.

Reply:
@plbossart @ujfalusi @andyross @marc-hb this PR is NOT about IPC ABI validation but a feature to support compliance/stress testing of kernel + SOF + Zephyr functional IP programming flows with userspace scripting (not userspace IP programming). 

We need to test the kernel functional IP code alongside the associated Zephyr IP code together (with the same timing and sequencing) which is not possible using Python to flip bits via mmap(). We already have a Python mmap() CI tool today that verifies IPC ABI compliance only. i.e. not useful to certify IP programming compliance. 

Yes we can do a lot of test with aplay/amixer, but these tools are end user applications not designed to strongly validate/stress the kernel + SOF + Zephyr combination.

e.g. we need to be able to stress (yes, low hanging fruit will come first)
1) D3 -> D0 and vice versa
2) DSP core off/on 
3) S0ix entry/exit
4) FW boot
5) IMR save/restore
6) topology load/unload

Reply:
I will assert that you can do all of the above with sof-test. Just run thousands of cycles of suspend-resume, module load/unload, etc.